core/src/main/scala/io/finch/EndpointResult.scala

```diff
@@ -60,69 +47,45 @@ sealed abstract class EndpointResult[+A] {
-   * is failed).
-   *
-   * @note This method is blocking. Never use it in production.
-   *
-   * @return `Some(output)` if this endpoint was matched on a given input, `None` otherwise.
-   */
-  final def awaitOutput(d: Duration = Duration.Top): Option[Try[Output[A]]] = this match {
-    case EndpointResult.Matched(_, _, out) => Some(Await.result(out.liftToTry.run, d))
+  def awaitOutput(d: Duration = Duration.Top)(implicit e: Effect[F]): Option[Try[Output[A]]] = this match {
+    case EndpointResult.Matched(_, _, out) =>
+      val (t, unit) = d.inTimeUnit
+      Some(Try(e.toIO(out).unsafeRunTimed(ScalaDuration(t, unit)) match {
+        case Some(a) => a
+        case _ => throw new TimeoutException(s"Output wasn't returned in time: $d")
+      }))
     case _ => None
   }
 
-  /**
-   * Awaits an [[Output]] of the [[Endpoint]] result or throws an exception if an underlying
-   * [[com.twitter.util.Future]] is failed.
-   *
-   * @note This method is blocking. Never use it in production.
-   *
-   * @return `Some(output)` if this endpoint was matched on a given input, `None` otherwise.
-   */
-  final def awaitOutputUnsafe(d: Duration = Duration.Top): Option[Output[A]] =
+  def awaitOutputUnsafe(d: Duration = Duration.Top)(implicit e: Effect[F]): Option[Output[A]] =
     awaitOutput(d).map(toa => toa.get)
 
-  /**
-   * Awaits a value from the [[Output]] wrapped with [[Try]] (indicating if either the
-   * [[com.twitter.util.Future]] is failed or [[Output]] wasn't a payload).
-   *
-   * @note This method is blocking. Never use it in production.
-   *
-   * @return `Some(value)` if this endpoint was matched on a given input, `None` otherwise.
-   */
-  final def awaitValue(d: Duration = Duration.Top): Option[Try[A]] =
+  def awaitValue(d: Duration = Duration.Top)(implicit e: Effect[F]): Option[Try[A]] =
     awaitOutput(d).map(toa => toa.flatMap(oa => Try(oa.value)))
 
-  /**
-   * Awaits a value from the [[Output]] or throws an exception if either an underlying
-   * [[com.twitter.util.Future]] is failed or [[Output]] wasn't a payload.
-   *
-   * @note @note This method is blocking. Never use it in production.
-   *
-   * @return `Some(value)` if this endpoint was matched on a given input,
-   *         `None` otherwise.
-   */
-  final def awaitValueUnsafe(d: Duration = Duration.Top): Option[A] =
+  def awaitValueUnsafe(d: Duration = Duration.Top)(implicit e: Effect[F]): Option[A] =
     awaitOutputUnsafe(d).map(oa => oa.value)
 }
 
 object EndpointResult {
 
-  final case class Matched[A](
+  final case class Matched[F[_], A](
     rem: Input,
     trc: Trace,
-    out: Rerunnable[Output[A]]
-  ) extends EndpointResult[A] {
+    out: F[Output[A]]
+  ) extends EndpointResult[F, A] {
     def isMatched: Boolean = true
   }
 
-  abstract class NotMatched extends EndpointResult[Nothing] {
+  abstract class NotMatched[F[_]] extends EndpointResult[F, Nothing] {
```

Comment:
Can we do following instead?

```scala
abstract class NotMatched extends EndpointResult[Pure, Nothing]
```

Also, as I asked bellow, can we use `Id` instead of `Pure`?

Reply:
> Can we do following instead? ...

It was my first intention, actually. Sadly, it's impossible due to the poor GADT support in Scala. Once you start to pattern-match `Endpoint.Result[F, A]` against `NotMatched extends EndpointResult[Pure, Nothing]` (and we do it a lot in `Endpoint`), everything turns into compile-errors nightmare and I spent around a week to find a work-around you see.

Other approach would be to write interpeters like [here](https://pchiusano.github.io/2014-05-20/scala-gadts.html), but it turns the nested pattern matching into spagetti.

> Can we use `Id` instead of `Pure`  

It's possible, though it's my extra precaution to avoid surprising implicit conversion of `EndpointResult[Id, ?]` into `EndpointResult[F[_], ?]` in customer codebase if users by some reasons use `Id` as an effect type. May be it's just my paranoia.

Reply:
Interesting. Let me play with it in a repl.